src/geodesy.jl

```diff
@@ -0,0 +1,447 @@
+The h-minima transformation suppresses all minima whose "relative" depth is below a given threshold (constant)
+# Arguments
+- `image::AbstractArray{T, N}`: where {T<:ImageCore.NumberLike , N} the ``N``-dimensional input image
+- `connectivity::AbstractArray{Bool}`: the neighborhood connectivity.
+- `constant::T`: the added constant.
+# Returns
+An array of the same type and shape as the `image`.
+# See also
+[`hmaxima`](@ref)
+# References
+    Morphological image analysis by Soille pg 170-172
+"""
+function hminima(image::AbstractArray{T, N}, connectivity::AbstractArray{Bool}, constant::T) where {T<:ImageCore.NumberLike , N}
+    tmp = image.+constant
+    return overbuild(tmp,image,connectivity)    
+end
+
+"""
+    underbuild(marker, mask, connectivity=[]) -> Array
+
+Morphological reconstruction by dilatation according to connection type
+
+# Details
+This operation performs dilations of marker restricted by the provided mask until idempotence This means that the output image
+will be included between mask and min(marker,mask).
+# Arguments
+- `marker::AbstractArray{T, N}`: where {T<:ImageCore.NumberLike , N} the ``N``-dimensional input image
+- `mask::AbstractArray{T, N}`: where {T<:ImageCore.NumberLike , N} the ``N``-dimensional input image
+- `connectivity::AbstractArray{Bool}`: the neighborhood connectivity.
+# Returns
+An array of the same type and shape as the `marker`.
+# See also
+[`overbuild`](@ref)
+# References
+  Morphological Grayscale Reconstruction in Image Analysis: Applications and Efficient Algorithms IEEE Trans Image Process. 1993;2(2)
+"""
+function underbuild(marker::AbstractArray{T, N}, mask::AbstractArray{T, N}, connectivity::AbstractArray{Bool}) where {T<:ImageCore.NumberLike , N}
+    check_image_consistency(mask, marker)
+    # WARNING
+    #  The original algorithm assume that the marker < mask, this assertion is not always true
+    #  so instead we extract the start from min(marker,mask)
+   
+    all(in((1,3)), size(connectivity)) || throw(ArgumentError("connectivity must have size 1 or 3 in each dimension"))
+    for d = 1:ndims(connectivity)
+        size(connectivity, d) == 1 || reverse(connectivity; dims=d) == connectivity || throw(ArgumentError("connectivity must be symmetric"))
+    end
+   
+    # first operation take the infimum between marker and mask
+    output = map(min, marker, mask)
+
+    queue = Queue{CartesianIndex{N}}(); 
+    deltaoffsets = neighborhood(marker, connectivity)
+    uper_deltaoffest = upper_half_neighborhood(marker, connectivity)
+    lower_deltaoffest= lower_half_neighborhood(marker, connectivity)
+
+    # forward scan
+    R = CartesianIndices(axes(marker))
+    for i in R
+        @inbounds current_max = output[i]
+        for Δi in uper_deltaoffest # examine neighborhoods
+            ii = i + Δi
+            if checkbounds(Bool, R, ii) #check that we are in the image
+                @inbounds current_max = max(current_max,output[ii])
+            end
+        end
```

Comment:
If we can pre-padding the `marker` with zeros (or `typemax` for overbuild), then we don't need to do `checkbounds` here, and thus can enable SIMD here or even utilizing LoopVectorization.

Reply:
Of course your right and its a general question, do we pad for almost all algorithm wich could benefit on it or the allocation cost and memory issues arise in 3D and we want to avoid padding ?
For exemple i thinking to write "generic" erode/dilate with arbitrary square SE, of course padding will make a huge gain :)

Reply:
I would say eagerly pad for almost all algorithms, this is also how ImageFiltering does. If we want to save memory, we can use virtual lazy padding, e.g., https://github.com/JuliaArrays/PaddedViews.jl

We don't currently have a flexible mechanism to switch between eager padding and lazy padding. This is actually part of ImageFiltering's 1.0 work to separate [`padarray`](https://github.com/JuliaImages/ImageFiltering.jl/blob/6f0dbc40637a763d6deff071056c8998bdb7a8b5/src/border.jl#L661-L665) out and potentially merge with PaddedViews.

Reply:
This can be redone in future PRs if you find it troublesome. But let's leave some comments as TODO note so that future contributors (maybe future you) know this.